Feature/file occurrences across prs

haydengomes opened the pull request:
Adding support for bitbucket server's pull-request diff endpoint. Additionally, this includes a new example which takes advantage of the endpoint to compute occurrences of files being modified across PRs to help identify churn as well as potential sources of bugs.

Commit:
Creating PR File occurrence counting example

- Added PR diff request with partial model implementation

Commit:
Adding comments to file occurrence example and reordering arguments

Commit:
Adding test for HttpBucketListClient.getPrDiff

aellett commented on bucketlist/src/main/kotlin/io/aexp/bucketlist/BucketListClient.kt:
```diff
@@ -34,6 +35,12 @@ interface BucketListClient {
     /**
+     * @return observable of PR diff response
+     */
+    fun getPrDiff(projectKey: String, repoSlug: String, prId: Long, contextLines: Int, whitespace: Boolean,
```
It's not really clear to me what contextLines, whitespace or withComments arguments are doing.  Maybe some more @params in the docs?

haydengomes replied:
These are params of the diff endpoint in the Stash REST API
https://developer.atlassian.com/static/rest/stash/3.11.3/stash-rest.html#idp1811216
I can add comments to the code clarifying their purpose based on the description in the API docs if you'd like

aellett commented on bucketlist/src/test/java/io/aexp/bucketlist/HttpBucketListClientTest.java:
```diff
@@ -178,6 +184,40 @@ public void testGetPrCommits() throws IOException {
+                .toBlocking()
+                .first();
+
+        assertEquals("16fb16e8afbe6c4087d39feddd68bdc881e302a2",diffResponse.getFromHash());
```
Looks like these asserts could use some spacing around the commas

haydengomes replied:
Fixed spacing in test

aellett commented on examples/src/main/kotlin/io/aexp/bucketlist/examples/prfileoccurrences/ExportPrFileOccurrencesData.kt:
```diff
@@ -0,0 +1,151 @@
+ * We wanted to better inform our decision of when to refactor different classes by identifying which were more highly
+ * correlated with bugs. So this determines in how many PRs across a given date range was each file changed. With the
+ * 'branch filter substring', we were able to reduce the set to only ones including our 'bugfix' prefix. The results are
+ * written
```
Hanging end of sentence here?

haydengomes replied:
Cleaned up comment

aellett commented on examples/src/main/kotlin/io/aexp/bucketlist/examples/prfileoccurrences/ExportPrFileOccurrencesData.kt:
```diff
@@ -0,0 +1,151 @@
+                .filter { pr -> pr.fromRef.id.contains(branchFilterSubstring, true) }
+                .map { pr -> pr.id }
+
+        val multiset: Multiset<String> = HashMultiset.create<String>()
```
You shouldn't need to specify the <String> here, right?

haydengomes replied:
Removed redundant second type argument

Commit:
Fixed spacing in testGetPrDiff and comment in ExportPrFileOccurrencesData example

Commit:
Updating params for getPrDiff

- Added params to interface comment
- Updated whitespace param to be a string value to match official API

yogurtearl commented:
lgtm

marshallpierce commented on bucketlist/src/main/kotlin/io/aexp/bucketlist/data/PullRequestDiff.kt:
```diff
@@ -0,0 +1,8 @@
+
+class PullRequestDiff(@JsonProperty("source") val source: PullRequestFilePath?,
+                      @JsonProperty("destination") val destination: PullRequestFilePath?,
+                      @JsonProperty("hunks") val hunks: List<PullRequestHunk>?,
```
When are these nullable?

marshallpierce commented on bucketlist/src/main/kotlin/io/aexp/bucketlist/BucketListClient.kt:
```diff
@@ -34,6 +35,17 @@ interface BucketListClient {
             Observable<PagedResponse<PullRequestCommit>>
 
     /**
+     * @param contextLines "the number of context lines to include around added/removed lines in the diff"
```
Unless something's changed in kdoc, I don't think we need the quotes

marshallpierce commented on bucketlist/src/main/kotlin/io/aexp/bucketlist/BucketListClient.kt:
```diff
@@ -34,6 +35,17 @@ interface BucketListClient {
 
     /**
+     * @param contextLines "the number of context lines to include around added/removed lines in the diff"
+     * @param whitespace "optional whitespace flag which can be set to 'ignore-all'" Default value is "show"
```
we should use enums rather than magic strings

marshallpierce commented on bucketlist/src/main/kotlin/io/aexp/bucketlist/BucketListClient.kt:
```diff
@@ -34,6 +35,17 @@ interface BucketListClient {
     /**
+     * @param contextLines "the number of context lines to include around added/removed lines in the diff"
+     * @param whitespace "optional whitespace flag which can be set to 'ignore-all'" Default value is "show"
+     * @param withComments "true to embed comments in the diff (the default); otherwise, false to stream the diff without comments"
```
consider a `CommentMode` enum or something to make a method call more readable: `foo.getPrDiff(..., true)` vs `foo.getPrDiff(..., WithComments)`. WDYT?

marshallpierce commented on bucketlist/src/test/java/io/aexp/bucketlist/HttpBucketListClientTest.java:
```diff
@@ -12,12 +12,19 @@
+import io.aexp.bucketlist.data.PullRequestDiff;
+import io.aexp.bucketlist.data.PullRequestDiffResponse;
+import io.aexp.bucketlist.data.PullRequestHunk;
 import io.aexp.bucketlist.data.PullRequestState;
```
can you juggle your formatting settings to not separate out the 'java' imports (to match existing style)

marshallpierce commented on examples/src/main/kotlin/io/aexp/bucketlist/examples/prfileoccurrences/ExportPrFileOccurrencesData.kt:
```diff
@@ -0,0 +1,150 @@
+ * correlated with bugs. So this determines in how many PRs across a given date range was each file changed. With the
+ * 'branch filter substring', we were able to reduce the set to only ones including our 'bugfix' prefix.
+ *
+ * This example is purely a data generator because we found that inspecting the top of the resulting list provided ample
```
I've never been terribly happy with "data generator" as a short description of "thing that outputs structured data instead of a pretty picture". I'm blanking on better terms, but still :(

marshallpierce commented on examples/src/main/kotlin/io/aexp/bucketlist/examples/prfileoccurrences/ExportPrFileOccurrencesData.kt:
```diff
@@ -0,0 +1,150 @@
+        val repoSlug = args[2]
+        val branchFilterSubstring = args[3]
+        val startDate = ZonedDateTime.of(LocalDateTime.of(LocalDate.parse(args[4]), LocalTime.MIDNIGHT), ZoneId.of("UTC"))
+        val endDate = ZonedDateTime.of(LocalDateTime.of(LocalDate.parse(args[5]), LocalTime.MIDNIGHT), ZoneId.of("UTC"))
```
this means the end date is exclusive. Perhaps you want 1 day after's midnight?

marshallpierce commented on examples/src/main/kotlin/io/aexp/bucketlist/examples/prfileoccurrences/ExportPrFileOccurrencesData.kt:
```diff
@@ -0,0 +1,150 @@
+                .first()
+
+        printToCsv(outputCsvPath, fileOccurrences)
+        System.exit(0)
```
What keeps the VM up if you don't exit? We should shutdown if possible

marshallpierce commented on bucketlist/src/main/kotlin/io/aexp/bucketlist/BucketListClient.kt:
```diff
@@ -34,6 +35,17 @@ interface BucketListClient {
+     * @param withComments "true to embed comments in the diff (the default); otherwise, false to stream the diff without comments"
+     * Param descriptions from official Atlassian documentation for Stash REST API
+     *
+     * @return observable of PR diff response
```
is there just one response object?

marshallpierce commented on examples/src/main/kotlin/io/aexp/bucketlist/examples/prfileoccurrences/ExportPrFileOccurrencesData.kt:
```diff
@@ -0,0 +1,150 @@
+        logger.debug("Fetching diff for PR " + prId);
+
+        return client.getPrDiff(projectKey, repoSlug, prId, 0, "ignore-all", false)
+                .first()
```
if `getPrDiff()` only returns one thing through its observable (which is unclear, but seems plausible), then `first()` has no effect

Commit:
Fixing comments, spacing, and date range

- Updated getPrDiff to clarify that its returned observable emits a single response
- Removed extra space around java imports
- Updated example to make the end date one day later than specified to be inclusive of the specified date rather than exclusive

Commit:
Converting withComment param to CommentMode enum

Commit:
Converting whitespace param to WhitespaceMode enum

Commit:
Adding clarity around PullRequestDiff params and their potential nullability

Commit:
Updated getBitBucketClient helper to take an AsyncHttpClient with default

- Updated ExportPrFileOccurrencesData example to pass in it's own AsyncHttpClient so it can be closed at the end of the process

Commit:
Fixing formatting on enum value names
